src/connectionbenchmark/benchmarktask.cpp

```diff
@@ -0,0 +1,119 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarktask.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+#include <QTimer>
+
+namespace {
+Logger logger(LOG_MAIN, "BenchmarkTask");
+}
+
+BenchmarkTask::BenchmarkTask(const BenchmarkType& benchmarkType,
+                             const uint32_t& maxRuntime)
+    : Task("BenchmarkTask"),
+      m_benchmarkType(benchmarkType),
+      m_maxRuntime(maxRuntime) {
+  MVPN_COUNT_CTOR(BenchmarkTask);
+}
+
+BenchmarkTask::~BenchmarkTask() { MVPN_COUNT_DTOR(BenchmarkTask); }
+
+void BenchmarkTask::setState(State state) {
+  logger.debug() << "Set state" << state;
+
+  m_state = state;
+}
+
+void BenchmarkTask::run() {
+  logger.debug() << "Run benchmark task";
+
+  if (m_state == StateCancelled) {
+    emit completed();
+    return;
+  }
+
+  setState(StateActive);
+  m_elapsedTimer.start();
+
+  connect(this, &BenchmarkTask::finished, [this]() {
+    setState(StateInactive);
+    emit completed();
+  });
+
```

Comment:
In order to have a better separation of classes, Add this protected method: `virtual void runInternal() = 0;`. Each implementation of `BenchmarkTask` will implement it.